src/operator/tensor/elemwise_binary_op.h

```diff
@@ -25,363 +25,795 @@
+
+    Tensor<cpu, 1, DType>   data_l = lhs.data().FlatTo1D<cpu, DType>(s);
+    Tensor<cpu, 1, DType>   data_r = rhs.data().FlatTo1D<cpu, DType>(s);
+    Tensor<cpu, 1, DType> data_out = output.data().FlatTo1D<cpu, DType>(s);
+
+    IType nnz = 0;
+    row_ptr_out[0] = 0;
+
+    for (IType i = 0; i < nr_rows; i++) {
+      IType head = -2;
+      IType length = 0;
+
+      // add a row of A to lhs_row
+      const IType i_start_l = row_ptr_l[i];
+      const IType i_end_l = row_ptr_l[i + 1];
+      for (IType jj = i_start_l; jj < i_end_l; jj++) {
+        IType col = col_indices_l[jj];
+        lhs_row[col] += data_l[jj];
+
+        if (next[col] == -1) {
+          next[col] = head;
+          head = col;
+          ++length;
+        }
+      }
+
+      // add a row of B to rhs_row
+      const IType i_start_r = row_ptr_r[i];
+      const IType i_end_r = row_ptr_r[i + 1];
+      for (IType jj = i_start_r; jj < i_end_r; jj++) {
+        const IType col = col_indices_r[jj];
+        rhs_row[col] += data_r[jj];
+
+        if (next[col] == -1) {
+          next[col] = head;
+          head = col;
+          ++length;
+        }
+      }
+
+      // scan through columns where A or B has
+      // contributed a non-zero entry
+      for (IType jj = 0; jj < length; jj++) {
+        const DType result = OP::Map(lhs_row[head], rhs_row[head]);
+
+        if (result != 0) {
+          col_indices_out[nnz] = head;
+          data_out[nnz] = result;
+          ++nnz;
+        }
+
+        const IType temp = head;
+        head = next[head];
+
+        next[temp] = -1;
+        lhs_row[temp] = 0;
+        rhs_row[temp] = 0;
+      }
+
+      row_ptr_out[i + 1] = nnz;
+    }
+  }
+
+  /*! \brief Minimum of three */
+  static MSHADOW_XINLINE size_t minthree(const size_t a, const size_t b, const size_t c) {
+    return a < b ? (a < c ? a : c) : (b < c ? b : c);
+  }
+
+  /*! \brief Maximum of three */
+  static MSHADOW_XINLINE size_t maxthree(const size_t a, const size_t b, const size_t c) {
+    return a > b ? (a > c ? a : c) : (b > c ? b : c);
+  }
+
+  /*! \brief LaunchEx allowing dense lvalue and/or rvalue */
+  template<typename xpu, typename OP, typename DType,
+    bool lhs_may_be_dense, bool rhs_may_be_dense, typename BackupCompute>
+  static void LaunchExDenseLRValue_(const nnvm::NodeAttrs &attrs,
+                                    const OpContext &ctx,
+                                    const std::vector<NDArray> &inputs,
+                                    const std::vector<OpReqType> &req,
+                                    const std::vector<NDArray> &outputs,
+                                    BackupCompute backup_compute) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(inputs.size(), 2);
+    CHECK_EQ(outputs.size(), 1);
+    if (req[0] != kNullOp) {
+      const NDArray *sparse = &inputs[0];
+      if (sparse->storage_type() == kDefaultStorage) {
```

Comment:
the conditionally-assigned 'sparse' variable is used later for RspRspElemwiseBinaryOp() in case MapToFCompute() is not called.